# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/canvas/WebGLContext.cpp
# Commit: 7c94d37c446e
# Full Hash: 7c94d37c446ea0d996bf6e6e04eb66101e126b5c
# Author: Jeff Gilbert <jgilbert@mozilla.com>
# Date: 2020-06-11 09:34:54
# Regressor Bug: 1632249
# File Overlap Count: 1
# Description:
#   Bug 1632249 - WebXR compositing fixes. r=imanol
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D78117
# ==============================================================================

diff -r 204b899f436d -r 7c94d37c446e dom/canvas/WebGLContext.cpp
--- a/dom/canvas/WebGLContext.cpp	Thu Jun 11 06:37:35 2020 +0000
+++ b/dom/canvas/WebGLContext.cpp	Thu Jun 11 06:37:36 2020 +0000
@@ -942,32 +942,46 @@
   return true;
 }
 
-void WebGLContext::Present() {
+void WebGLContext::Present(WebGLFramebuffer* const fb,
+                           const layers::TextureType consumerType) {
   const FuncScope funcScope(*this, "<Present>");
   if (IsContextLost()) return;
 
-  mResolvedDefaultFB = nullptr;
+  auto swapChain = &mSwapChain;
+  if (fb) {
+    swapChain = &fb->mOpaqueSwapChain;
+  } else {
+    mResolvedDefaultFB = nullptr;
+  }
 
-  if (!mSwapChain.mFactory) {
+  if (!swapChain->mFactory) {
+    auto typedFactory = gl::SurfaceFactory::Create(gl, consumerType);
+    if (typedFactory) {
+      swapChain->mFactory = std::move(typedFactory);
+    }
+  }
+  if (!swapChain->mFactory) {
     NS_WARNING("Failed to make an ideal SurfaceFactory.");
-    mSwapChain.mFactory = MakeUnique<gl::SurfaceFactory_Basic>(*gl);
+    swapChain->mFactory = MakeUnique<gl::SurfaceFactory_Basic>(*gl);
   }
-  MOZ_ASSERT(mSwapChain.mFactory);
+  MOZ_ASSERT(swapChain->mFactory);
 
-  (void)PresentInto(mSwapChain);
+  (void)PresentInto(*swapChain);
 }
 
 Maybe<layers::SurfaceDescriptor> WebGLContext::GetFrontBuffer(
-    const layers::TextureType consumerType) {
-  if (mSwapChain.mFactory &&
-      mSwapChain.mFactory->mDesc.consumerType != consumerType) {
-    auto typedFactory = gl::SurfaceFactory::Create(gl, consumerType);
-    if (typedFactory) {
-      mSwapChain.mFactory = std::move(typedFactory);
-    }
+    WebGLFramebuffer* const fb, const layers::TextureType consumerType) {
+  auto swapChain = &mSwapChain;
+  if (fb) {
+    swapChain = &fb->mOpaqueSwapChain;
   }
 
-  const auto& front = mSwapChain.FrontBuffer();
+  if (swapChain->mFactory &&
+      swapChain->mFactory->mDesc.consumerType != consumerType) {
+    swapChain->mFactory = nullptr;  // Better luck next Present.
+  }
+
+  const auto& front = swapChain->FrontBuffer();
   if (!front) return {};
 
   return front->ToSurfaceDescriptor();